### PR TITLE
[Core] Use sys.executable in shim if conda not specified

### LINKED ISSUE #15807: [RuntimeEnv] The setup_runtime_env.py choose incorrect Python
### What is the problem?

The `setup_runtime_env.py` execvp `bash -c exec python ...`, while the default Python may not be the Python ray is using. For example, system default Python is Python 2.7,  ray use Python 3.6.5. The Python worker will fail.

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray version: Latest master 395c0ea03c5e90e2df91071606f9a1444068c189
system default Python 2.7
ray Python 3.8.2

The error log of ray_client_server.err
``` log
INFO:ray.util.client.server.server:Starting Ray Client server on 0.0.0.0:10003
INFO:ray.util.client.server.logservicer:New logs connection established. Total clients: 1
2021-05-14 13:03:10,467	INFO worker.py:715 -- Connecting to existing Ray cluster at address: 10.15.239.87:6379
[2021-05-14 13:03:10,475 I 7866 6533027] logging.cc:288: Set ray log level from environment variable RAY_BACKEND_LOG_LEVEL to -1
INFO:filelock:Lock 4600493104 acquired on /tmp/ray/session_2021-05-14_13-03-06_575483_7758/runtime_resources/_ray_pkg_5ae4f99d9c00ae48531f9494639cd7f5.zip.lock
INFO:filelock:Lock 4600493104 released on /tmp/ray/session_2021-05-14_13-03-06_575483_7758/runtime_resources/_ray_pkg_5ae4f99d9c00ae48531f9494639cd7f5.zip.lock
2021-05-14 13:03:10,655	WARNING worker.py:1102 -- Failed to unpickle the remote function '__main__.run_test' with function ID 8e418a68251f80073c2166d1f9161562aa530e58a84d787ed9ceea04. Traceback:
Traceback (most recent call last):
  File "/Users/po/Work/ray-ant/python/ray/_private/function_manager.py", line 180, in fetch_and_register_remote_function
    function = pickle.loads(serialized_function)
  File "/Users/po/Work/ray-ant/python/ray/cloudpickle/cloudpickle.py", line 584, in subimport
    __import__(name)
ModuleNotFoundError: No module named 'test_module'

[2m[33m(raylet)[0m   File "/Users/po/Work/ray-ant/python/ray/workers/default_worker.py", line 107
[2m[33m(raylet)[0m     f"{ray_constants.LOGGING_ROTATE_BYTES} bytes.")
[2m[33m(raylet)[0m     ^
[2m[33m(raylet)[0m SyntaxError: invalid syntax
2021-05-14 13:03:28,442	WARNING worker.py:1102 -- The actor or task with ID f28b92927ed74e6fffffffffffffffffffffffff02000000 cannot be scheduled right now. It requires {CPU: 1.000000} for placement, but this node only has remaining {1.000000/1.000000 CPU, 7.337038 GiB/7.337038 GiB memory, 0.146484 GiB/0.146484 GiB object_store_memory, 1.000000/1.000000 node:10.15.239.87}
. In total there are 1 pending tasks and 0 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):
test_runtime_env.py::test_single_node

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Closes https://github.com/ray-project/ray/issues/15807

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/workers/setup_runtime_env.py [hunk 1]
```diff
@@ -20,11 +21,15 @@ def setup(input_args):
     commands = []
     runtime_env: RuntimeEnvDict = json.loads(args.serialized_runtime_env
                                              or "{}")
+
+    py_executable: str = sys.executable
+
     if runtime_env.get("conda"):
         if isinstance(runtime_env["conda"], str):
             commands += get_conda_activate_commands(runtime_env["conda"])
+            py_executable = "python"
```
Comment (architkulkarni) [inline python/ray/workers/setup_runtime_env.py:30]: When a conda environment is activated, `python`will point to the correct executable for that conda environment.

Comment (fyrestone) [reply python/ray/workers/setup_runtime_env.py:30]: Conda activate will make the bash to find the correct Python. But, It's dangerous to run bash with user inputs. For example, user may write a special conda name to remove all data from all nodes.

Comment (architkulkarni) [reply python/ray/workers/setup_runtime_env.py:30]: Ah thanks, that's a really good point.  I wonder what would be a good way to validate the conda name input. One thought I had was to simply run "conda activate <name>" in a subprocess and check if it fails, but that suffers from the same problem you mentioned.  I made an issue for it here https://github.com/ray-project/ray/issues/15837

**Review by edoakes — APPROVED** [review]

File: python/ray/workers/setup_runtime_env.py (hunk 1 above)
Comment (fyrestone) [reply python/ray/workers/setup_runtime_env.py:30]: > Ah thanks, that's a really good point. I wonder what would be a good way to validate the conda name input. One thought I had was to simply run "conda activate " in a subprocess and check if it fails, but that suffers from the same problem you mentioned. I made an issue for it here #15837

This PR use the `pydantic` to validate the RuntimeEnv. https://github.com/ray-project/ray/pull/15804

**Pull request merged**